clients/src/main/java/org/apache/kafka/clients/consumer/internals/GroupCoordinator.java

```diff
@@ -0,0 +1,635 @@
+        this.coordinator = null;
+        this.sessionTimeoutMs = sessionTimeoutMs;
+        this.heartbeat = new Heartbeat(this.sessionTimeoutMs, heartbeatIntervalMs, time.milliseconds());
+        this.heartbeatTask = new HeartbeatTask();
+        this.sensors = new GroupCoordinatorMetrics(metrics, metricGrpPrefix, metricTags);
+        this.requestTimeoutMs = requestTimeoutMs;
+        this.retryBackoffMs = retryBackoffMs;
+    }
+
+    /**
+     * Get the current list of protocols (and their associated metadata) supported
+     * by the local member.
+     * @return Non-empty list of supported protocols
+     */
+    protected abstract M metadata();
+
+    /**
+     * Get the sub-protocols supported by this member. This will be used by the coordinator to verify
+     * compatibility among members
+     * @return A non-empty list of protocol names which this member supports
+     */
+    protected abstract Collection<String> subProtocols();
+
+    /**
+     * Invoked when a new group is joined.
+     * @param generation The generation that was joined
+     * @param memberId The identifier for the local member in the group.
+     * @param memberState
+     */
+    protected abstract void onJoin(int generation, String memberId, S memberState);
+
+    /**
+     * Perform synchronization for the group. This is used by the leader to push state to all the members
+     * of the group (i.e. to push partition assignments in the case of the new consumer)
+     * @param leaderId The id of the leader (which is this member)
+     * @param allMemberMetadata Metadata from all members of the group
+     * @return A map from each member to their assignment
+     */
+    protected abstract Map<String, S> doSync(String leaderId, String subProtocol, Map<String, M> allMemberMetadata);
+
+
+    /**
+     * Invoked when the group is left (whether because of shutdown, metadata change, stale generation, etc.)
+     * @param generation The generation that was left
+     * @param memberId The identifier of the local member in the group
+     */
+    protected abstract void onLeave(int generation, String memberId);
+
+
+    /**
+     * Block until the coordinator for this group is known.
+     */
+    public void ensureCoordinatorKnown() {
+        while (coordinatorUnknown()) {
+            RequestFuture<Void> future = sendGroupMetadataRequest();
+            client.poll(future, requestTimeoutMs);
+
+            if (future.failed())
+                client.awaitMetadataUpdate();
+        }
+    }
+
+    /**
+     * Check whether the group should be rejoined (e.g. if metadata changes)
+     * @return true if it should, false otherwise
+     */
+    public boolean needRejoin() {
+        return rejoinNeeded;
+    }
+
+    /**
+     * Ensure that the group is active (i.e. joined and synced)
+     */
+    public void ensureActiveGroup() {
+        if (!needRejoin())
+            return;
+
+        // onLeave only invoked if we have a valid current generation
```

Comment:
Seems like this comment is not true now? Intentional change or are we missing a check here?